File: go-kata/internal/data-structures/array_list/array_list_test.go
package data_structures_test

import (
	array_list "go-kata/internal/data-structures/array_list"
	"testing"

	"github.com/stretchr/testify/require"
)

func createArrayList(capacity int, args ...int) *array_list.ArrayList[int] {
	arr := array_list.NewIntArrayList(capacity)

	for _, v := range args {
		arr.Push(v)
	}

	return arr
}

func TestArrayListPush(t *testing.T) {
	arr := createArrayList(3, 69, 44)
	firstVal, err := arr.Get(0)
	require.Nil(t, err, "Problems with getting from array")
	firstValNeg, err := arr.Get(-2)
	require.Nil(t, err, "Problems with getting from array")
	lastVal, err := arr.Get(1)
	require.Nil(t, err, "Problems with getting from array")
	lastValNeg, err := arr.Get(-1)
	require.Nil(t, err, "Problems with getting from array")

	require.Equal(t, arr.GetLength(), 2, "Array length should equal 2")
	require.Equal(t, arr.GetCapacity(), 3, "Array capacity should equal 3")
	require.Equal(t, 69, firstVal, "The first value should be 69")
	require.Equal(t, 69, firstValNeg, "The first value (by negative indx) should be 69")
	require.Equal(t, 44, lastVal, "The last value should be 44")
	require.Equal(t, 44, lastValNeg, "The last value (with negative indx) should be 44")
}

File: typescript-kata/src/__tests__/data-structures/array-list.test.ts
import { ArrayList } from "src/data-structures/array-list";
import { describe, expect, test } from "vitest";

function createArrayList(capacity: number, ...args: number[]): ArrayList {
    const arrayList = new ArrayList(capacity)

    for (const arg of args) {
        arrayList.push(arg)
    }

    return arrayList
}

describe("Array List test suite", () => {
    test("Empty Array List should have length of 0", () => {
        const arr = createArrayList(3)
        expect(arr.getLength()).toEqual(0)
    })

    test("Empty Array List should have capacity of 3", () => {
        const arr = createArrayList(3)
        expect(arr.getCapacity()).toEqual(3)
    })

    test("After pushing till capacity, capacity should equal length", () => {
        const arr = createArrayList(3, 69, 42, 97)
        expect(arr.getCapacity()).toEqual(arr.getLength())
    })

    test("After pushing over capacity, capacity should double", () => {
        const arr = createArrayList(3, 69, 42, 97, 44)
        expect(arr.getCapacity()).toEqual(6)
    })

    test("Popping should return 97", () => {
        const arr = createArrayList(3, 69, 42, 97)
        expect(arr.pop()).toEqual(97)
    })

    test("The last element should be 42", () => {
        const arr = createArrayList(3, 69, 42) 

        expect(arr.get(1)).toEqual(42)
        expect(arr.get(-1)).toEqual(42)
    })

    test("The first element should be 69", () => {
        const arr = createArrayList(3, 69, 42) 

        expect(arr.get(-2)).toEqual(69)
        expect(arr.get(0)).toEqual(69)
    })


    test("Getting element out of range should return undefined", () => {
        const arr = createArrayList(3, 69, 42) 

        expect(arr.get(-3)).toEqual(undefined)
        expect(arr.get(2)).toEqual(undefined)
    })
})
